lib/LanguageServerWorseReflection/Workspace/WorkspaceIndex.php

```diff
@@ -37,41 +46,63 @@ public function documentForName(Name $name): ?TextDocument
 
     public function index(TextDocument $textDocument): void
     {
+        $this->documents[(string)$textDocument->uri()] = $textDocument;
+        $this->updateDocument($textDocument);
+    }
+
+    private function updateDocument(TextDocument $textDocument): void
+    {
+        $newNames = [];
         foreach ($this->reflector->reflectClassesIn($textDocument) as $reflectionClass) {
-            $this->byName[$reflectionClass->name()->full()] = $textDocument;
+            $newNames[] = $reflectionClass->name()->full();
         }
-
+        
         foreach ($this->reflector->reflectFunctionsIn($textDocument) as $reflectionFunction) {
-            $this->byName[$reflectionFunction->name()->full()] = $textDocument;
+            $newNames[] = $reflectionFunction->name()->full();
         }
+
+        $this->updateNames($textDocument, $newNames, $this->documentToNameMap[(string)$textDocument->uri()] ?? []);
     }
 
-    public function update(TextDocumentUri $textDocumentUri, string $updatedText): void
+    private function updateNames(TextDocument $textDocument, array $newNames, array $currentNames): void
     {
-        foreach ($this->byName as $className => $textDocument) {
-            if ($textDocumentUri != $textDocument->uri()) {
-                continue;
-            }
+        $namesToRemove = array_diff($currentNames, $newNames);
+        
+        foreach($newNames as $name) {
+            $this->byName[$name] = $textDocument;
+        }
+        foreach($namesToRemove as $name){
+            unset($this->byName[$name]);
+        }
+
+        if(!empty($newNames))
+            $this->documentToNameMap[(string)$textDocument->uri()] = $newNames;
+        else
+            unset($this->documentToNameMap[(string)$textDocument->uri()]);
```

Comment:
cs fixer is failing, probably this is one of them, just run `./vendor/bin/php-cs-fixer fix` (or `composer integrate`)